time_utils/automation/lib/python/community/time_utils.py

```diff
@@ -108,45 +115,99 @@ def to_datetime(when, log=logging.getLogger("{}.time_utils".format(LOG_PREFIX)))
+            dt_java = to_java_zoneddatetime(dt_python)
+
         elif isinstance(when, datetime):
-            dt = to_joda_datetime(when)
+            log.debug('when is datetime')
+            dt_python = when
+            dt_java = to_java_zoneddatetime(when)
+            #get system time zone to avoid convertion errors
+            dt = to_joda_datetime(when.replace(tzinfo=None)) 
+
+        elif isinstance(when, time):
+            log.debug('when is python time object')
+            dt_java = ZonedDateTime.now().withHour(when.hour).withMinute(when.minute).withSecond(when.second)
+            dt_python = to_python_datetime(dt_java)
+            dt = to_joda_datetime(dt_java)
+
+        elif isinstance(when, (str, unicode)):
+            if is_iso8601(when):
+                log.info('when is iso8601: '+str(when))
+                dt = DateTime(when)
+                dt_python = datetime.strptime(str(when), "%Y-%m-%dT%H:%M:%S.%fffff%Z")
+                #get system time zone to avoid convertion errors
+                dt_java = to_java_zoneddatetime(dt_python.replace(tzinfo=None))
+            else:
+                log.debug('when is duration')
+                log.debug(str(when))
+                dt = parse_duration_to_datetime(when, log)
+                log.debug('dt is ' + str(dt))
+                dt_python = datetime.now() + parse_duration(when, log)
+                log.debug('dt python is ' + str(dt))
+                #get system time zone to avoid convertion errors
+                dt_java = to_java_zoneddatetime(dt_python.replace(tzinfo=None))
         else:
-            log.warn("When is an unknown type {}".format(type(when)))
+            log.warn('When is an unknown type {}'.format(type(when)))
     except:
-        import traceback
-        log.error("Exception: {}".format(traceback.format_exc()))
-    finally:
+        log.error('Exception: {}'.format(traceback.format_exc()))
+
+    if output == 'python':
+        log.debug('returning dt python')
+        return dt_python
+    elif output == 'Java':
+        log.debug("returning dt java")
+        return dt_java
+    else:
+        log.debug("returning dt joda")
         return dt
 
-def to_today(when, log=logging.getLogger("{}.time_utils".format(LOG_PREFIX))):
+def to_today(when, log=logging.getLogger("{}.time_utils".format(LOG_PREFIX)), output='Joda'):
     """Takes a when (see to_datetime) and updates the date to today.
     Arguments:
         - when : One of the types or formats supported by to_datetime
         - log: optional logger, when not supplied one is created for logging errors
     Returns:
         - DateTime specified by when with today's date.
     """
+    log.debug(output)
 
-    dt = to_datetime(when, log)
-    now = dt.now()
+    if output == 'python':
+        dt = to_datetime(when, log=log, output = 'python')
+        now = datetime.now()
+        return dt.replace(year=now.year, month=now.month, day=now.day)
```

Comment:
Wouldn't it be better to use

    dt = datetime.date.fromtimestamp(dt.time())

to update the Python datetime to today? There is less conversion back and forth that way.

Reply:
Wouldn't that change the time when going to and from daylight saving?
I think time zone information would be lost also

Reply:
I admit I've never tested this but time() uses epoch, which is all that a datetime is at it's core anyway. When you call methods like `localtime()` is when stuff like timezone and DLS gets applied to the epoch. So I would expect this to work as expected. Though where trouble occurs is when we have a mix of Aware and Naive datetimes. In that case it might not work. 

And now that I look at the above, that line probably won't do what I expect anyway. But something like the following should.

    dt = datetime.combine(date.today(), dt.time())

It will grab the default TZ from the environment. And since we are not doing an offset but setting the time explicitly the DST changeover would be handled too.